### ISSUE TITLE
Allow parameters to make request unique

### ISSUE THREAD

**Issue by tomkuijsten** [tomkuijsten/restup#49]
jQuery has a cache parameter when doing ajax calls. It adds a "_" query parameter. This breaks the REST uri match and you will receive a bad request response. It should somehow be possible to allow this kind of parameters, but still match your uri.

**Comment by Jark** [comment]
Related to https://github.com/tomkuijsten/restup/issues/35 ?

**Comment by tomkuijsten** [comment]
Jee :) It probably has some effect on #35, but they are different in functionality.

**Comment by Jark** [comment]
@tomkuijsten

Good to see that you guys are still using that term :) I'm trying to introduce it here too ;)

That's true, but also very related. You could solve both by writing a url query parser and use that in the matching algorithm.

Let's say a method has the following query uri pattern:

```
[UriFormat("/?param1={param1}&param2={param2}")]
public IGetResponse DoAction(string param1, int param2) {
}
```

When parsing the method definitions we could store it like this:
```
var params = new List<Params>();
params.Add(new Param("param1", typeof(string))); // or something generics based
params.Add(new Param("param2", typeof(int))); // or something generics based
```

Make sure the rest method definitions are sorted by most params first.

Then let's say the following comes in:
```
GET /?param1=hello&param2=123&_=asldkfjslak
```

We can then again parse this as:

```
var params = new List<QueryParams>();
params.Add(new QueryParam("param1")); // or something generics based
params.Add(new QueryParam("param2")); // or something generics based
params.Add(new QueryParam("_")); // or something generics based
```

And then when we loop through the rest definitions to find the first match, have the match method be:

```
function bool IsMatch() {
if (!regex path matches)  // matching that happens now on everything except the query path
   return false;

return UriFormatParams.All(x => IncomingQueryParams.Contains(x.Name));
```

As a bonus we can reuse this to make implementing #16 as well.

Anyway, just a thought :)

**Comment by tomkuijsten** [comment]
Good thoughts :). The problem with ignoring parameters is the shown in this sample. Imagine uri pattern 1:

```cs 
[UriFormat("/?param1={param1}&param2={param2}")]
```

And pattern 2:

```cs 
[UriFormat("/?param1={param1}&param3={param3}")]
```

And i query for

```
http://server/api/?param1=1&param2=2&param3=3
```

Which pattern should it match?

**Comment by Jark** [comment]
Hey @tomkuijsten,

Sorry for the late reply, have been quite busy lately.

I think in that case the api creator could expect undefined behaviour.

If the api creator decides to define the exact same paths (remember it matches on prefix as well, empty in this case) with the same named parameters, and then an api consumer calls it with more parameters than the calls have, then I think undefined behaviour is what you can expect.

What do you think?

**Comment by tomkuijsten** [comment]
I agree with the strict matching. So do we want to allow the user the choose the query parameters which should be ignored? So let's say I want to ignore "_" as a query parameter, I should add it when configuring the webserver, eg:

```cs
var webserver  = new HttpServer();
webserver.Configuration.IgnorableParameters = new[] {"_"};
```

**Comment by Jark** [comment]
I read http://api.jquery.com/jquery.ajax/ and apparently the _={timestamp} is only used for GET requests (which makes sense, since only GET responses should be cached).

What I meant with my reply is that if you have the following UriFormats:

`[UriFormat("/?param1={param1}&param2={param2}")]`

`[UriFormat("/?param1={param1}&param3={param3}")]`

and then this call comes in:

`http://server/api/?param1=1&param2=2&param3=3&_=123123`

It will match on whatever one is first in the list. (so in this case the first one) The undefined behaviour in this case is that the writer of the call should know which rest method he's trying to call. 

I think in practice this situation will be an edge case, since if you follow proper REST design you wouldn't have overlapping parameters imho. 

I like to avoid configuration where possible until the situation arises, it reduces complexity in this case I think.

**Comment by tomkuijsten** [comment]
I agree and think we should just implement it the way you suggested.

**Comment by tomkuijsten** [comment]
I ended up implementing it in a slightly different way. Too prevent confusing matches of uri's, I held on to the idea that the given ```UriFormat``` should match exactly (no different order of parameters allowed). The only thing I added is the option for extra query parameters, which are ignored if there not part of the ```UriFormat```.

@Jark care for another review? :)

**Comment by tomkuijsten** [comment]
Btw, I didn't clean up the code yet, only made all tests work again.

**Comment by stevegg** [comment]
The way this is commonly done in Java Frameworks (Jersey and its derivatives) is that urls are matched but parameters are not.  Missing parameters are given default values.  (e.g. strings are "", numeric values are 0).  You use an attribute on the parameters in the method signature to match the query parameter and (optionally) to set  default values.  For example:

        @Path("/foo")
        public Response listItems(
	        @DefaultValue("0") @QueryParam("start") int start,
	        @DefaultValue("25")@QueryParam("to") int count ) {


In C# I could see it looking something like this:

       [UriFormat("/foo")]
       public IGetResponse ListItems( [QueryParam("start", "0")] int start, 
                                      [QueryParam("count", "25")int count )

**Comment by Jark** [comment]
@tomkuijsten I'll have a look at your modified code today, is it still ready for code review?

**Comment by tomkuijsten** [comment]
Let me put it this way... there is room for improvement. Feels a bit too hacky now. I do feel for an approach as suggested by @stevegg . Or at least make a separation between the url segments and query parameters.

**Comment by Jark** [comment]
@tomkuijsten I'll take a look then and try and make some improvements on top of your code, we can then see if those are improvements or not :)

**Comment by Jark** [comment]
@tomkuijsten I did a bit of prototyping work around parsing the urls differently which I think will make for a solid base, but haven't finished yet. Will hopefully be able to push something up today.

**Comment by tomkuijsten** [comment]
Would be nice :) Want to release a stable version any time "soon"...

**Comment by Jark** [comment]
Hi @tomkuijsten,

Have a look at https://github.com/tomkuijsten/restup/compare/development...Jark:parameters-code-review everything is building and unit tests are green

I've added a https://github.com/Jark/restup/blob/d382a5ba16fcdaaebac62dc4dadf2e78cb019e3d/src/WebServer/Rest/UriParser.cs which parses any string uri into its relative parts (https://github.com/Jark/restup/blob/d382a5ba16fcdaaebac62dc4dadf2e78cb019e3d/src/WebServer/Rest/ParsedUri.cs).

UriMatches now works like this: https://github.com/Jark/restup/blob/d382a5ba16fcdaaebac62dc4dadf2e78cb019e3d/src/WebServer/Rest/RestControllerMethodInfo.cs#L182

And getting the parameters from the uri now works like this:

https://github.com/Jark/restup/blob/d382a5ba16fcdaaebac62dc4dadf2e78cb019e3d/src/WebServer/Rest/RestControllerMethodInfo.cs#L204

There's still some complexity I'd like to remove around _pathParameterNameToIndex  / _queryParameterNamesByMethodName (with polymorphism we should be able to combine them) and in parsing the uri twice (e.g. UriMatches should return a MatchResult with the ParsedUri or something and pass that in the method invoke), but for the most part I think it works quite nicely.

Let me know what you think about this approach.

Cheers,

Jark

**Comment by Jark** [comment]
Hey @tomkuijsten,

I'm done with the refactoring. Let me know if you have any questions / don't think this is a good idea :) 

Cheers,

Jark

**Comment by tomkuijsten** [comment]
PR?

**Comment by tomkuijsten** [comment]
Fixed with #73
